### PR TITLE
Adding task for publishing test attestation metadata to evidence store

### PR DESCRIPTION
**Pull request by dabasnidhi**

### CODE REVIEW

File: Tasks/PublishPipelineMetadataV0/publishmetadata.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                if (result.statusCode < 200 && result.statusCode >= 300) {
+                    throw tl.debug("publishToImageMetadataStore failed with error: " + result.statusMessage);
+                }
+            }, (error) => {
+                throw tl.debug("publishToImageMetadataStore failed with error: " + error);
+            });
+        });
+    }
+    catch (error) {
+        tl.setResult(tl.TaskResult.Failed, error);
```
Comment (vithati) [inline Tasks/PublishPipelineMetadataV0/publishmetadata.ts]: log warning and mark it as success. This should not fail the job.

File: Tasks/PublishPipelineMetadataV0/publishmetadata.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            return;
+        }
+
+        const requestUrl = tl.getVariable("System.TeamFoundationCollectionUri") + tl.getVariable("System.TeamProject") + "/_apis/deployment/attestationdetails?api-version=5.2-preview.1";
+        metadataObjects.forEach((requestPayload: any) => {
+            const requestObject: AttestationRequestPayload = constructMetadataRequestBody(requestPayload);
+            sendRequestToImageStore(JSON.stringify(requestObject), requestUrl).then((result) => {
+                tl.debug("ImageDetailsApiResponse: " + JSON.stringify(result));
+                if (result.statusCode < 200 && result.statusCode >= 300) {
+                    throw tl.debug("publishToImageMetadataStore failed with error: " + result.statusMessage);
```
Comment (vithati) [inline Tasks/PublishPipelineMetadataV0/publishmetadata.ts]: don't throw . log it and proceed futher.

File: Tasks/PublishPipelineMetadataV0/publishmetadata.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    humanReadableName?: string;
+    serializedPayload: string;
+}
+
+function getPipelineMetadataObjects(): any {
+    const pipelineMetadataRequestBodyString = "METADATA";
+    const allVariables = tl.getVariables();
+    let requestObjects = [];
+    allVariables.forEach(v => {
+        if (v.name.startsWith(pipelineMetadataRequestBodyString)) {
```
[the remark this answers is not in the document]
Comment (dabasnidhi) [reply Tasks/PublishPipelineMetadataV0/publishmetadata.ts]: Changed both var name and metadata string to lower case before comparison

File: Tasks/PublishPipelineMetadataV0/publishmetadata.ts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    const pipelineMetadataRequestBodyString = "metadata";
+    const allVariables = tl.getVariables();
+    let requestObjects = [];
+    allVariables.forEach(v => {
+        if (v.name.toLowerCase().startsWith(pipelineMetadataRequestBodyString)) {
+            try {
+                requestObjects.push(JSON.parse(v.value));
+            }
+            catch (error) {
+                tl.debug("Failed to parse metadata variable; Error: " + error);
```
Comment (vithati) [inline Tasks/PublishPipelineMetadataV0/publishmetadata.ts]: add v.name to exception.

File: Tasks/PublishPipelineMetadataV0/publishmetadata.ts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        const requestUrl = tl.getVariable("System.TeamFoundationCollectionUri") + tl.getVariable("System.TeamProject") + "/_apis/deployment/attestationdetails?api-version=5.2-preview.1";
+        metadataObjects.forEach((requestPayload: any) => {
+            const requestObject: AttestationRequestPayload = constructMetadataRequestBody(requestPayload);
+            sendRequestToImageStore(JSON.stringify(requestObject), requestUrl).then((result) => {
+                tl.debug("ImageDetailsApiResponse: " + JSON.stringify(result));
+                if (result.statusCode < 200 && result.statusCode >= 300) {
+                    tl.debug("publishToImageMetadataStore failed with error: " + result.statusMessage);
+                }
+            }, (error) => {
+                tl.debug("publishToImageMetadataStore failed with error: " + error);
```
Comment (vithati) [inline Tasks/PublishPipelineMetadataV0/publishmetadata.ts]: add request payload.

File: Tasks/PublishPipelineMetadataV0/publishmetadata.ts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        if (metadataObjects.length == 0) {
+            tl.debug("Not pushing as no metadata found");
+            return;
+        }
+
+        const requestUrl = tl.getVariable("System.TeamFoundationCollectionUri") + tl.getVariable("System.TeamProject") + "/_apis/deployment/attestationdetails?api-version=5.2-preview.1";
+        metadataObjects.forEach((requestPayload: any) => {
+            const requestObject: AttestationRequestPayload = constructMetadataRequestBody(requestPayload);
+            sendRequestToImageStore(JSON.stringify(requestObject), requestUrl).then((result) => {
+                tl.debug("ImageDetailsApiResponse: " + JSON.stringify(result));
```
Comment (vithati) [inline Tasks/PublishPipelineMetadataV0/publishmetadata.ts]: add requestobject to the debug log.

Comment (dabasnidhi) [reply Tasks/PublishPipelineMetadataV0/publishmetadata.ts]: We log the request object inside sendRequestToImageStore function itself, so it is not required to log here again

**Pull request merged**